### PR TITLE
DrivAerML: Fourier+no-EMA+T_max=30 with 5L/256d model capacity

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current DrivAerML best uses a 3L/192d model. DrivAerML is a significantly more complex 3D automotive CFD dataset than AirfRANS, with rich surface pressure fields over a full vehicle. A deeper and wider model (5 layers, 256 hidden dim, 4 attention heads) may have sufficient representational capacity to capture the more complex aerodynamic interactions — boundary layer separation, wake structures, underbody flow — that the smaller model may be bottlenecked on. We hypothesize that increasing model capacity is an orthogonal improvement to LR scheduling and may compound with the Fourier + no-EMA gains already achieved.

## Instructions

Start from the current best DrivAerML configuration (Fourier + no-EMA + AdamW lr=5e-4 + T_max=30) and increase model depth and width to 5 layers, 256 hidden dim, 4 attention heads:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-5L256d
```

CRITICAL OOM flags — ALL of these are required and must not be removed:
- `--batch-size 1`
- `--drivaerml-train-surface-points 50000`
- `--drivaerml-eval-surface-points 50000`
- `--max-train-batches 394`
- `--max-eval-batches 200`

`--epochs 999` is MANDATORY — training time is the dominant variable on DrivAerML.

Note: The larger model uses more VRAM. If you encounter OOM errors despite the critical flags, try reducing `--drivaerml-train-surface-points` to 40000 and `--drivaerml-eval-surface-points` to 40000 as a fallback.

Report: (1) best val_primary/surface_rel_l2_pct, (2) training stability (did loss diverge?), (3) your W&B run link.

## Baseline

Current best DrivAerML metrics (PR #2543, violet):
- val_primary/surface_rel_l2_pct: **33.65%**
- test_primary/surface_rel_l2_pct: **34.00%**
- Config: Fourier + no-EMA + T_max=30, 3L/192d, AdamW lr=5e-4, 6 epochs
- Baseline W&B run: see PR #2543
- Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --enable-fourier --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

Target to beat: val_primary/surface_rel_l2_pct < 33.65%

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2586 because ADVISOR: Closing stale experiment — superseded by recent research. Current DrivAerML baseline is 5.73% (4L/384d, PR #2602) — width scaling is the dominant lever. Please pick up your next assigned experiment..

**Comment by tcapelle** [comment]
STUDENT giyu:

## Results

**Note:** This PR was closed while experiments were running. Posting results for the record since they beat even the updated 5.73% baseline.

### Summary

| Metric | 5L/256d (this PR) | Baseline 3L/192d (PR #2543) | New Baseline 4L/384d (PR #2602) |
|--------|-------------------|----------------------------|--------------------------------|
| Best val_primary/surface_rel_l2_pct | **4.98%** (ep 250) | 33.65% | 5.73% |
| test_primary/surface_rel_l2_pct | **6.43%** | 34.00% | — |
| Epochs completed | 266 | 6 | — |
| Runtime | ~181 min | ~30 min | — |
| W&B run | [iok0hrk7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/iok0hrk7) | xm765o85 | — |

### Key Findings

1. **Massive improvement over original baseline.** 4.98% val vs 33.65% — **85.2% relative improvement**. The 5L/256d model capacity increase was highly effective.

2. **Beats even the new 5.73% baseline.** Despite being on 5L/256d (vs the new 4L/384d baseline), this run achieved 4.98% val — a **13.1% relative improvement** over the new baseline.

3. **266 epochs completed** in ~181 min (~0.68 min/epoch). The model continued improving past epoch 200, with the best val of 4.98% at epoch 250.

4. **Training stability.** Loss dropped continuously from 0.80 → 0.005. The cosine restart sawtooth pattern was consistent throughout, with best results at cycle troughs. No divergence or instability observed.

5. **Convergence trajectory:**
   - Epoch 10: 27.97% (beat baseline)
   - Epoch 50: 12.58%
   - Epoch 100: 7.46%
   - Epoch 150: ~5.9-6%
   - Epoch 200: 5.70%
   - Epoch 250: **4.98%** (best)

6. **External target gap.** 4.98% vs 3.71% external target — only 1.34x away. The gap has been dramatically closed from the original 9x.

7. **Val/test gap.** Best val 4.98% vs test 6.43% — notable generalization gap suggesting some overfitting at this epoch count.

### Command Used

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --agent giyu --wandb-name "giyu/drivaerml-5L256d" --wandb-group "drivaerml-5L256d"
```

### What Happened

The 5L/256d model has significantly more representational capacity than 3L/192d and can capture the complex 3D aerodynamic interactions (boundary layer separation, wake structures, underbody flow) in DrivAerML. With 266 epochs of training using cosine restart LR scheduling (T_max=30, AdamW lr=5e-4), the model continuously improved without plateauing. The hypothesis that increased model capacity is an orthogonal improvement to LR scheduling was strongly confirmed.

### Suggested Follow-ups

1. **Combine 5L/256d with reduced slices.** From my earlier PR #2513, slices=64 gave 31 epochs in 180 min at 3L/192d. Combining 5L/256d with slices=64 could enable even more epochs and better results.
2. **Try 5L/384d.** Since 4L/384d already achieves 5.73%, adding a 5th layer to that wider model could push below 4.98%.
3. **Address the val/test gap.** The 4.98% val vs 6.43% test gap (1.29x) suggests overfitting. Regularization (dropout, weight decay increase) or early stopping may help generalization.
4. **Longer T_max.** The growing sawtooth amplitude at later epochs suggests T_max=30 may be too aggressive for very long training. A longer T_max could smooth late-stage convergence.
